Improve crate selection on rustdoc search results page

steffahn opened the pull request:
Improves some problems pointed out in #93240.

In particular the `in` being serif-font makes sense once this is in a new line, and the font-size is better so the dropdown menu isn’t this huge anymore. This also addresses my concern that the selection moves around horizontally as the search term changes.

The formatting for the `#search-settings` CSS is not necessary for the title, because `h1` has the same CSS, too, anyways.

Currently on nightly:
![Screenshot_20220703_200911](https://user-images.githubusercontent.com/3986214/177052139-85f7a506-00c9-4c93-a84c-3cec4a7b7ee8.png)
![Screenshot_20220703_200924](https://user-images.githubusercontent.com/3986214/177052136-ccb2f3c2-4c2c-4348-a7aa-44e0fda72a47.png)


~~After this PR:~~ _(outdated, see below)_
![Screenshot_20220703_201333](https://user-images.githubusercontent.com/3986214/177052213-e75c7087-6667-42e0-b003-980d944908f4.png)

![Screenshot_20220703_200829](https://user-images.githubusercontent.com/3986214/177052157-010c832c-7778-4ad2-a209-8655845d347e.png)

---

With my latest changes added, I'd say this PR can close #93240 entirely. See https://github.com/rust-lang/rust/pull/98855#issuecomment-1176822718 below to see renderings of the current status.

steffahn committed:
Improve crate selection on rustdoc search results page

Resolves all of issue #93240

Reproduces a similar change as #99086, but with improvements

In particular, this PR inlcludes:
* redesigning the crate-search selector so the background color matches its surroundings
* decrease the font of the dropdown menu to a reaonable size
* add a hover effect
* make the color of the arrow theme-dependent, using a surrounding div, with :after pseudo-element
  that can then be transformed using CSS filters to approximate the desired color
* fix the text "in" to match the title font
* remove the "for xyz" in the "Results for xyz in [All crates]" title when
  searching for search term "xyz"; you can already see what you're searching for
  as it's typed in the search bar!
* in line with #99086, handle super-long crate names appropriately without a long <select>
  element escaping the screen area; the improvement is that we also keep the title
  within a single line now; uses some flex layout shenanigans...
* the margins / paddings are adjusted so the selected label of the <select> fits within
  the rest of that title nicely; also some inconsistency in the way that Firefox renders
  a <select> with "appearance: none" (roughly 4px more padding left and right of the text
  than e.g. Chrome) is worked around, and it now produces a result that looks (essentially)
  identical to Chrome
* the color of the help menu and settings menu border in light theme is made to match with
  the color of the corresponding buttons, like they do (match) in the ayu theme
* the casing of "All crates" changes to "all crates"
* the new tests from #99086 are temporarily disabled, until they can be adapted later

rust-highfive commented:
r? @jsha

(rust-highfive has picked a reviewer for you, use r? to override)

GuillaumeGomez commented:
I uploaded an online version of this change [here](https://rustdoc.crud.net/imperio/improve_rustdoc_search_results_page_crates_selection/foo/index.html?search=foo). I'm personally not a big fan of putting the "in X" part in its own line but let's see what others think about it.

cc @jsha @Manishearth

steffahn commented:
The thing you uploaded only has a single crate, so the list doesn’t show up in the first place. It did however remind me that in that case, the `<br/>` probably shouldn’t appear in the HTML (even though it doesn’t appear to get rendered as a new line anyway), so after re-inspecting the logic in the JS, I’ve moved it into the `crates` variable to make it conditional like that.

jsha commented:
I agree with @GuillaumeGomez that we shouldn't use a second line of vertical space. @steffahn you commented in #93240 that you are indifferent to whether the results page echoes the search query. I think that might be a better solution than using a second line; let's remove the "for `search_term`" part. Then the dropdown can be on the same line but not move around as you edit your search.

steffahn commented:
Okay, keeping it in the title and removing the search term; making the `in` part of the title font (but not the selection, that looks weird in bold). I would still like to keep the selection dropdown smaller, if that’s stylistically acceptable. So how about something like this for a look? (I just manually edited this in the browser so far, not implemented it yet into the source files.)

![Screenshot_20220706_192234](https://user-images.githubusercontent.com/3986214/177608520-09ab7963-d78e-4296-92c4-0c5cfbac09e6.png)

With a single crate, the thing would *just* say Results.

steffahn commented:
On second thought, in e.g. the white theme where the dropdown is lower-contrast, keeping matching fonts seems kind-of essential actually. Compare:

![Screenshot_20220706_193911](https://user-images.githubusercontent.com/3986214/177610814-a5f215db-018e-405b-8d17-a90b3da7c644.png)

![Screenshot_20220706_193924](https://user-images.githubusercontent.com/3986214/177610830-57d8b509-2500-4ac1-9687-d7c9371ecc8d.png)

the second looks better IMO.

steffahn commented:
> theme where the dropdown is lower-contrast

Makes me reconsider perhaps also addressing the other point I made in #93240 here, the high contrast in the dark theme currently. Is there any other dropdown-selections in rustdoc that should be taken into consideration for consistency?

steffahn commented:
I added an implementation of a style along those lines, i.e. as explained in my previous two comments.

![Screenshot_20220707_003120](https://user-images.githubusercontent.com/3986214/177654536-a1df2ec2-0234-458e-95c5-8e560fe7508b.png)
![Screenshot_20220707_003155](https://user-images.githubusercontent.com/3986214/177654544-7c439792-b442-4ff5-a307-f04ab1f73e2d.png)
![Screenshot_20220707_003229](https://user-images.githubusercontent.com/3986214/177654552-b4358597-aeb5-4e63-9d4d-3914f8a8fc97.png)

Small font size in dropdown:
![Screenshot_20220707_003414](https://user-images.githubusercontent.com/3986214/177654556-72b5c1ff-2370-4bb4-95dc-297278efa9f2.png)

As you might be able to tell by looking into the CSS, I had a hard time adapting the color of that arrow to the field; I’m open for suggestions for cleaner approaches if an inline svg like this is not acceptable.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/7224037890?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/5d0f22d120a272b2524ce47b329a029ff19a4ec9/checks/7224037890/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
* highest error code: E0788
Found 505 error codes
Found 0 error(s) in error codes
Done!
tidy error: /checkout/src/librustdoc/html/static/css/themes/light.css:199: line longer than 100 chars
tidy error: /checkout/src/librustdoc/html/static/css/themes/ayu.css:231: line longer than 100 chars
tidy error: /checkout/src/librustdoc/html/static/css/themes/dark.css:208: line longer than 100 chars
some tidy checks failed
Build completed unsuccessfully in 0:00:10
```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/7224264557?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/b488ae8e1dd299beab68cc0faf6f3dee1e9e5e15/checks/7224264557/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
* highest error code: E0788
Found 505 error codes
Found 0 error(s) in error codes
Done!
tidy error: /checkout/src/librustdoc/html/static/css/themes/light.css:201: CSS files use tabs for indent
tidy error: /checkout/src/librustdoc/html/static/css/themes/light.css:203: CSS files use tabs for indent
tidy error: /checkout/src/librustdoc/html/static/css/themes/ayu.css:233: CSS files use tabs for indent
tidy error: /checkout/src/librustdoc/html/static/css/themes/ayu.css:235: CSS files use tabs for indent
tidy error: /checkout/src/librustdoc/html/static/css/themes/dark.css:210: CSS files use tabs for indent
tidy error: /checkout/src/librustdoc/html/static/css/themes/dark.css:212: CSS files use tabs for indent
some tidy checks failed
Build completed unsuccessfully in 0:00:15
```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/7224671038?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/f6787abb6a20ecaf0ff252254ce036167e27e162/checks/7224671038/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
info: ES-Check: there were no ES version matching errors!  🎉
+ eslint -c ../src/librustdoc/html/static/.eslintrc.js ../src/librustdoc/html/static/js/externs.js ../src/librustdoc/html/static/js/main.js ../src/librustdoc/html/static/js/scrape-examples.js ../src/librustdoc/html/static/js/search.js ../src/librustdoc/html/static/js/settings.js ../src/librustdoc/html/static/js/source-script.js ../src/librustdoc/html/static/js/storage.js

/checkout/src/librustdoc/html/static/js/search.js
  1554:14  error  'escape' is defined but never used. Allowed unused vars must match /^_/u  no-unused-vars

✖ 1 problem (1 error, 0 warnings)
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/7224843141?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/0f6a9c96414feaa0532f42877bf6341925867046/checks/7224843141/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
  IMAGE: x86_64-gnu-tools
##[endgroup]
From https://github.com/rust-lang/rust
 * branch              master     -> FETCH_HEAD
Searching for toolstate changes between 7665c3543079ebc3710b676d0fd6951bedfd4b29 and c2e979b7221877cdecaf03ea028dfcbacc9cefda
Rustdoc was updated
##[group]Run src/ci/scripts/verify-channel.sh
src/ci/scripts/verify-channel.sh
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
env:
---
.......... (60/66)
....     (66/66)


/checkout/src/test/rustdoc-gui/pocket-menu.goml pocket-menu... FAILED
[ERROR] (line 72) Error: Evaluation failed: The following errors happened (for selector `#help-button .popover`): [expected `rgb(221, 221, 221)` for key `border-color`, found `rgb(224, 224, 224)`]: for command `assert-css: (
    "#help-button .popover",
    {"display": "block", "border-color": "rgb(221, 221, 221)"},


/checkout/src/test/rustdoc-gui/search-filter.goml search-filter... FAILED
[ERROR] (line 41) Error: Evaluation failed: property `value` (`all crates`) isn't equal to `All crates` for selector `#crate-search`: for command `assert-property: ("#crate-search", {"value": "All crates"})`
[ERROR] (line 51) "#search-settings" not found: for command `assert-text: ("#search-settings", "Results in all crates", STARTS_WITH)`
[ERROR] (line 70) Error: The following CSS properties still don't match: [border: (`1px solid rgb(210, 210, 210)` != `1px solid rgb(240, 240, 240)`), color: (`rgb(221, 221, 221)` != `rgb(17, 17, 17)`), background-color: (`rgb(53, 53, 53)` != `rgb(240, 240, 240)`)]: for command `wait-for-css: ("#crate-search", {
    "border": "1px solid rgb(240, 240, 240)",
    "color": "rgb(17, 17, 17)",
    "background-color": "rgb(240, 240, 240)",


/checkout/src/test/rustdoc-gui/search-result-display.goml search-result-display... FAILED
[ERROR] (line 5) Error: The following CSS selector "#search-settings" was not found: for command `wait-for: "#search-settings"`
Build completed unsuccessfully in 0:00:43
```

</details>

GuillaumeGomez commented on src/librustdoc/html/static/js/search.js:
```diff
@@ -1710,22 +1704,15 @@ function initSearch(rawSearchIndex) {
         if (crates_list.length > 1) {
-            crates = " in <select id=\"crate-search\"><option value=\"All crates\">" +
-                "All crates</option>";
+            crates = " in <select id=\"crate-search\"><option value=\"all crates\">" +
```
Why this change?

steffahn replied:
because "Result in all crates" reads better than "Result in All crates". Or do you mean the internal representation changing from "All crates" to "all crates", too? That just felt more consistent than just changing the label, but if there's a reason against it, it can of course be kept "All crates" internally.

GuillaumeGomez replied:
I just checked locally and I think it actually renders quite nicely locally.

steffahn replied:
By the way, feel free to do your magic again for uploading an online versions of this change for others to check out.

GuillaumeGomez replied:
I'll after your next update. I really like this change. :)

GuillaumeGomez commented on src/librustdoc/html/static/css/themes/ayu.css:
```diff
@@ -220,10 +220,22 @@ details.undocumented > summary::before {
 	/* Without the `!important`, the border-color is ignored for `<select>`... */
-	border-color: #424c57 !important;
+	border-color: #5c6773 !important;
+	background-image: url('data:image/svg+xml;utf8,<svg xmlns="http://www.w3.or\
```
I gave a try to a svg minifier (just in case) and the result is indeed a bit shorter mostly because of the `<svg>` attributes:

```
data:image/svg+xml;utf8,<svg width="128" height="128" enable-background="new 0 0 128 128" version="1.1" viewBox="-30 -20 176 176" xml:space="preserve" xmlns="http://www.w3.org/2000/svg"><g fill="none" stroke="%235c6773" stroke-linecap="square" stroke-miterlimit="10" stroke-width="12"><line x1="111" x2="64" y1="40.5" y2="87.499"/><line x1="64" x2="17" y1="87.499" y2="40.5"/></g></svg>
```

GuillaumeGomez replied:
Just realized but please put `background-image` into `rustdoc.css` to prevent copying it into multiple places.

steffahn replied:
It's a different svg depending on the theme (the color is different). That's the reason I've put it inline in the first place, since despite some searching, I couldn't find a nice way to modify the color of the existing svg from CSS. I just copied the pre-existing file `down-arrow.svg` and modified the color. I'm open to better solutions. E. g. to use `filter`, which can change color, one would somehow need to make the arrow it's own element or pseudo-element, presumably involving some helper-`div`. I don't know, perhaps I could give such a restructure a try, but I'm really not all that familiar with html and css and this stuff in the first place, so it's quite a bit of effort.

steffahn replied:
The status quo is particularly unclean because of the forced line breaks because the tidy check otherwise fails the CIt.

GuillaumeGomez replied:
Ah I see. I had to do that recently so I can maybe help you with that. Give me a few minutes. :smile:

GuillaumeGomez replied:
You can disable the tidy check on a line iirc. To be confirmed.

GuillaumeGomez replied:
I see. Sorry, misunderstood what you meant previously. Then yes, you need to use a pseudo element.

steffahn replied:
In the latest commit, I use a filter like that now, with a `:after` pseudo-element and a `div`. I hope everything works (or at least is not in a badly broken state) in all browsers we want to support; I would appreciate if you were able to create an online version again so I (and perhaps others) can test this on more devices.

GuillaumeGomez commented:
Just realized that if you don't need to use `background-image` in all themes, we already have it in `rustdoc.css`.

GuillaumeGomez commented:
That seems very weird to switch the "focus" from the main content to the sidebar, no?

steffahn commented:
I have yet to find out why the thing has so much more visual padding in the `<select>` on Firefox than Chrome, and how to avoid this. (I prefer the way it looks on Chrome / that’s what I looked at when changing/choosing the padding in the first place.)

Chrome
![Screenshot_20220709_000525](https://user-images.githubusercontent.com/3986214/178076156-047deb83-5869-4f93-b3a3-0c94f8af4ae1.png)

Firefox
![Screenshot_20220709_000546](https://user-images.githubusercontent.com/3986214/178076172-49248e92-9651-4452-a81d-5070556af035.png)

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/7258905373?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/ad8451830883a65b30bae3df81d734c0ec800e73/checks/7258905373/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
* highest error code: E0788
Found 505 error codes
Found 0 error(s) in error codes
Done!
tidy error: /checkout/src/librustdoc/html/static/css/rustdoc.css:921: CSS files use tabs for indent
tidy error: /checkout/src/librustdoc/html/static/css/rustdoc.css:944: CSS files use tabs for indent
some tidy checks failed
Build completed unsuccessfully in 0:00:11
```

</details>

steffahn commented:
Some screenshots of the hover effects I’ve added yesterday:
![Screenshot_20220709_094014](https://user-images.githubusercontent.com/3986214/178096788-1315aadf-fa8f-44e9-92cc-8820bf9dfca0.png)
![Screenshot_20220709_094209](https://user-images.githubusercontent.com/3986214/178096791-8972d5b7-d479-4961-b012-79af25d33b9d.png)
![Screenshot_20220709_094105](https://user-images.githubusercontent.com/3986214/178096794-4200705f-fe47-4451-b02a-91b54a574cd8.png)
![Screenshot_20220709_094142](https://user-images.githubusercontent.com/3986214/178096796-3c5cffdc-d3bc-4303-b7d9-12546fcd5f90.png)
![Screenshot_20220709_093903](https://user-images.githubusercontent.com/3986214/178096798-e6f83ddb-6615-4f40-befb-29e558415717.png)
![Screenshot_20220709_094236](https://user-images.githubusercontent.com/3986214/178096799-d161f558-e0ac-4e48-9d4a-8188e1cd6f78.png)

For light and ayu, colors of outline (both with and without focus, respectively) are the same as the settings and help button; the border color without focus is also the same as the outline of the settings *menu* on all themes, so that’s how I chose the outline color for the dark theme; the blue hover color of the dark theme is the same as the hover color for the radio buttons and the activate switches in the settings menu, admitted there are a lot of slightly different shades of blue, i.e. the searchbar highlight above (if it’s focused) is a different blue, and the active tab below (“In Names”/“In Parameters”/“In Return Types”) has yet-another shade of blue.

GuillaumeGomez commented:
I uploaded the docs generated with this PR [here](https://rustdoc.crud.net/imperio/improve_rustdoc_search_results_page_crates_selection/foo/index.html?search=test).

bors commented:
:umbrella: The latest upstream changes (presumably #99112) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

steffahn commented:
@GuillaumeGomez

I’m going to resolve the merge conflict by improving upon #99086 by keeping the title in a single line
![Screenshot_20220710_231838](https://user-images.githubusercontent.com/3986214/178162431-23e14edd-a6ea-4030-a525-21e72d1aeaaf.png)

---

Next up, I’ll also squash the commits and rebase; the current history is mostly me fighting with CI or completely re-designing everything, anyway.

steffahn commented:
@GuillaumeGomez, I have no idea whatsoever how these `.goml` tests work in detail, and how I need to adjust / re-write the test to test for the new improved behavior I’ve implemented. So I feel like I’m unable to make this pass CI on my own. I would appreciate some help, e.g. explanation of what these tests even mean / what *exactly* they test, line-by-line. (I’m referring to the ones introduced in #99086.) Or alternatively, maybe you could simply write an appropriate updated test for me? Either posted here, or perhaps I’d disable the test and you can create a new one in a separate PR?

(It doesn’t help that last time I tried, I had trouble getting these tests to run on my machine in the first place, and running them with CI takes about half an hour until I get feedback.)

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/7273798816?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/be4ef1272940dc938cc2dd673286037bba42d96e/checks/7273798816/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
  IMAGE: x86_64-gnu-tools
##[endgroup]
From https://github.com/rust-lang/rust
 * branch              master     -> FETCH_HEAD
Searching for toolstate changes between c396bb3b8a16b1f2762b7c6078dc3e023f6a2493 and bfc6f26ecca8133bfd22e604c14e8d580bafb6b6
Rustdoc was updated
##[group]Run src/ci/scripts/verify-channel.sh
src/ci/scripts/verify-channel.sh
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
env:
---
.......... (60/66)
......     (66/66)


/checkout/src/test/rustdoc-gui/search-result-display.goml search-result-display... FAILED
[ERROR] (line 19) Error: Evaluation failed: The following errors happened (for selector `#crate-search`): [expected `222px` for key `width`, found `223.797px` (or `224px`)]: for command `assert-css: ("#crate-search", {"width": "222px"})`
[ERROR] (line 20) "#search-settings .search-results-title" not found: for command `compare-elements-position-near: (
    "#crate-search",
    "#search-settings .search-results-title",
    {"y": 5},
)`
[ERROR] (line 33) Error: Evaluation failed: The following errors happened (for selector `#crate-search`): [expected `640px` for key `width`, found `526.828px` (or `527px`)]: for command `assert-css: ("#crate-search", {"width": "640px"})`
[ERROR] (line 34) "#search-settings .search-results-title" not found: for command `compare-elements-position-near-false: (
    "#crate-search",
    "#search-settings .search-results-title",
    {"y": 5},

Build completed unsuccessfully in 0:00:22
```

</details>

steffahn commented:
I’ve disabled the abovementioned failing (still rather new) tests for now; if it’s okay to have them only come back in a future PR, then this PR should be ready to be reviewed and merged from my end, provided CI passes now 🚀

GuillaumeGomez commented on src/librustdoc/html/static/css/rustdoc.css:
```diff
@@ -934,14 +949,35 @@ table,
 	text-indent: 0.01px;
-	text-overflow: "";
+}
+/* cancel stylistic differences in padding
```
When does it happen? We don't use `appearance: none` ourselves after all.

steffahn replied:
I think we effectively are, kind-of… I guess… at least with `appearance: none` it looks the same, IDK about backwards-compatibility with older versions of Firefox where it might’ve been different..?

In any case, what this is supposed to say is that Firefox has some (empirically, 4px) padding in an "unstyled" <select> (i.e. when the native arrow is removed) that Chrome doesn't have. I don't have any other browsers myself ATM to test / compare.

Admitted, given that `appearance: none` is **not** literally what’s used here, the comment could be clarified.

steffahn replied:
I changed the wording a bit now.

I’m not sure about the pros and cons of using the “actual” `appearance: none` property, I haven’t researched too much what distinguishes this property from those `-moz-appearance` / `-webkit-appearance` ones; consequently I kept everything as-is with regards to these properties, just to be safe, e.g. since I haven’t tested other browsers, etc…

GuillaumeGomez commented on src/librustdoc/html/static/css/themes/dark.css:
```diff
@@ -196,11 +196,26 @@ details.undocumented > summary::before {
+}
+#crate-search:hover, #crate-search:focus {
+	border-color: #2196f3 !important;
+	border-color: #008dfd !important;
```
Something weird is going on here. :)

steffahn replied:
Looks like some copy-paste error, oops…

I’ll double-check which color is correct.

GuillaumeGomez commented on src/test/rustdoc-gui/search-result-display.goml:
```diff
@@ -15,26 +15,28 @@ assert-css: (".search-results div.desc", {"width": "570px"})
-    "#search-settings .search-results-title",
-    {"y": 5},
-)
+// FIXME: Fix and re-enable these tests!
```
So about the `goml` format. It's something I wrote to make it easy to write/read. You can see more information about it (with the full list of commands) [here](https://github.com/GuillaumeGomez/browser-UI-test/blob/master/goml-script.md#command-list).

steffahn committed:
Two small improvements:

Fix oversight duplicate property left in CSS (dark theme).

Improve wording in comment that mentions `appearance: none`

steffahn commented:
> We can take a look at the `goml` tests at the end if you're still stuck. Not really the hard part normally. ;)

Well for one thing, I’d need to get those tests running locally in the first place, otherwise it’s impossible to work with.

It just gives me a completely cryptic error message
```
Running 66 rustdoc-gui (8 concurrently) ...
ERROR: puppeteer failed when trying to create a new page. Please try again with `--no-sandbox`
```

I’m also noticing this warning on the `npm install browser-ui-test`:
```
npm WARN deprecated puppeteer@3.3.0: Version no longer supported. Upgrade to @latest
```

I have no idea if that’s related to the problem and if yes how to fix it.

steffahn commented:
W.r.t. rewriting/adapting the tests, they currently (judging by the comments) seem to work with the implicit assumption that the `<select>` lives in its own line and just test for its size. What’s *actually* interesting is presumably that its right-hand-side border is positioned within-bounds of, say, the `search` section, I have no idea how one would write this.

I also don’t know what this `compare-elements-position-near`-stuff means or what the error message talking about `"#search-settings .search-results-title"` means, in particular what the `"#search-settings .search-results-title"` syntax itself means. Looks like two CSS selectors separated by space; IDK how those compose.

steffahn commented:
TBH, I don’t have much more time I can spend on this today. I would highly appreciate if I didn’t have to learn this whole testing infrastructure now, too. This PR is already so much more that I initially wanted to do. Learning new HTML and CSS fundamentals I didn’t know before is enough commitment for my liking, I don’t feel like going too deep into this whole testing framework, too, now.

GuillaumeGomez commented:
I can do this part then.

bors commented:
:umbrella: The latest upstream changes (presumably #99346) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

GuillaumeGomez commented:
I took the commits from here, fixed the conflicts and added the missing GUI tests in https://github.com/rust-lang/rust/pull/100374.

steffahn commented:
Looks like at some point Firefox must have changed, at least my Firefox browser no longer shows the differences that used to be compensated with the
```css
/* cancel stylistic differences in padding in firefox
for "appearance: none"-style (or equivalent) <select>s */
@-moz-document url-prefix() {
	#crate-search {
		padding-left: 0px; /* == 4px - 4px */
		padding-right: 19px; /* == 23px - 4px */
	}
}
```
part, hence that can (and should) probably be removed. I’ll need to look into this more, and create an appropriate new issue and PR if this is really the case. (Makes me wonder whether there are good ways to do something like “display a website for a bunch of different versions of Firefox” in order to confirm that there really was a change at some version throughout the last year, and I’m not the one that changed a factor.

GuillaumeGomez commented:
I was wondering the same thing a few days ago. Don't hesitate to send a PR in any case!

As for checking on different firefox versions, I remember that websites allow it, but don't remember the name. ^^'

steffahn commented:
I didn’t find any websites easily. I’ll just ignore the urge to conclusively “prove” something in Firefox has changed, and just go with the observation that both on my PC and phone, the current Firefox seems to be displaying it differently than intended and removing the special case workaround CSS code seems like the way to fix it. So I’ll just open a PR tomorrow. :-)

GuillaumeGomez commented:
Perfect, thanks!
